csrc/multidevice/communication.cpp

```diff
@@ -249,13 +249,11 @@ c10::intrusive_ptr<c10d::Work> postAllgather(
     c10d::Backend* backend,
     at::Tensor input_tensor,
     at::Tensor output_tensor) {
-  std::vector<at::Tensor> input_tensors({input_tensor});
-  std::vector<std::vector<at::Tensor>> output_tensors(1);
-  output_tensors[0] = at::split(output_tensor, /*split_size=*/1, /*dim=*/0);
+  auto splits = at::split(output_tensor, /*split_size=*/1, /*dim=*/0);
+  assertBufferCount(splits, communication->team().size());
+  assertBuffersHaveSameSize({input_tensor}, splits);
 
-  assertBufferCount(output_tensors[0], communication->team().size());
-  assertBuffersHaveSameSize(input_tensors, output_tensors[0]);
-  return backend->allgather(output_tensors, input_tensors, {});
+  return backend->_allgather_base(output_tensor, input_tensor, {});
```

Comment:
```suggestion
  // allgather primitive in c10d induces extra buffering time to copy out the received tensors into user buffer. It is therefore always preferable to use _allgather_base, which does not perform any extra copy at the cost of assuming that the receive buffers are placed contiguously.
  // See #2384 for an illustration. 
  return backend->_allgather_base(output_tensor, input_tensor, {});
```

Reply:
It'd also be helpful to link to the PyTorch code where this extra buffering happens. I'm a bit surprised that the default API is slow and the fix seems to be a quite straightforward replacement. cc @Aidyn-A and @syed-ahmed to see if it's worth a fix upstream.

Reply:
> It'd also be helpful to link to the PyTorch code where this extra buffering happens.

The extra copy happens: 
- in nccl: [here](https://github.com/pytorch/pytorch/blob/1edcb31d34ef012d828bb9f39a8aef6020f580b2/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp#L3589
) 
- in ucc: [here](https://github.com/pytorch/pytorch/blob/1edcb31d34ef012d828bb9f39a8aef6020f580b2/torch/csrc/distributed/c10d/ProcessGroupUCC.cpp#L940). Note that in ucc there is an alternative solution, which is [to use allgatherv](https://github.com/pytorch/pytorch/blob/1edcb31d34ef012d828bb9f39a8aef6020f580b2/torch/csrc/distributed/c10d/ProcessGroupUCC.cpp#L873) with a destination buffer equalling `nullptr`; this avoids the extra copy, however, allgatherv is less efficient than allgather

> I'm a bit surprised that the default API is slow and the fix seems to be a quite straightforward replacement. cc @Aidyn-A and @syed-ahmed to see if it's worth a fix upstream.

I am not sure this deserves a "fix" upstream, it is done like that on purpose, to provide more flexibility to the user.
1) On the one hand, communication backend like ucc and nccl (idk for Gloo) work with raw void pointers, i.e., fully contiguous src and dst buffers. This restriction is necessary for performance.
2) On the other hand, ProcessGroup API allows users to provide a vector of Tensors instead of a raw pointer for one-to-many, many-to-many and many-to-one collective operations. This makes the API very convenient for users who don't need advanced optimization. In other words, ProcessGroup::allgather expects as dst buffers a vector of n buffer of size m, whereas nccl and ucc expect one buffer of size n*m. To bridge this gap, ProcessGroup needs to copy in/out the vector of buffers into one big contiguous chunk before using the backend, hence the extra cudaMemcpyAsync


To provide "best of both worlds" to the advanced users, ProcessGroup exposes `_allgather_base` `_reduce_scatter_base` and `alltoall_base`, which directly use nccl or ucc backend with raw pointers assuming the buffers are fully contiguous.

The other "base" collectives (e.g. scatter_base, gather_base) are missing in the API though.

Therefore, imo, there is no change upstream necessary, however, as far as nvFuser is concerned, we should only use the "base" collective when they are available.

Coming next a PR to patch `reduce_scatter`